finality-aleph/src/sync/substrate/verification/mod.rs

```diff
@@ -94,4 +129,52 @@ where
             },
         }
     }
+
+    fn verify_header(&mut self, mut header: Header) -> Result<Header, Self::Error> {
+        use HeaderVerificationError::*;
+        // compare genesis header directly to the one we know
+        if header.number().is_zero() {
+            return match &header == self.genesis_header() {
+                true => Ok(header),
+                false => Err(Self::Error::HeaderVerification(IncorrectGenesis)),
+            };
+        }
+        // off by one!
+        let parent_number = header.number() - 1;
+        let slot = find_pre_digest::<Block, AuthoritySignature>(&header)
+            .map_err(|e| Self::Error::HeaderVerification(PreDigestLookupError(e)))?;
+        // duplicate code, watch out!
+        let slot_now = Slot::from_timestamp(
+            sp_timestamp::Timestamp::current(),
+            sp_consensus_slots::SlotDuration::from_millis(MILLISECS_PER_BLOCK),
+        );
+        // timorl's offset
+        if slot > slot_now + 10 {
+            return Err(Self::Error::HeaderVerification(HeaderTooNew(slot)));
+        }
+        // pop the seal BEFORE hashing
+        let seal = header
+            .digest_mut()
+            .pop()
+            .ok_or(Self::Error::HeaderVerification(MissingSeal))?;
+        let sig = seal
+            .as_aura_seal()
+            .ok_or(Self::Error::HeaderVerification(IncorrectSeal))?;
+        let authorities = self
+            .get_aura_authorities(parent_number)
+            .map_err(|_| Self::Error::HeaderVerification(MissingAuthorityData))?;
+        // duplicate code, watch out!
```

Comment:
This is a bit cryptic, duplicate with what exactly?

Reply:
Do you mean "assuming round robin"? That's hardly an explanation. <_<

Reply:
could you suggest something more descriptive then?

Reply:
Unfortunately I have no idea what this (or actually most of the comments about duplication in this function) means – that's why I'm asking this question. xd

If it's only about duplication with Aura, then perhaps just having a comment above the function about this that details either the differences or the duplication (whichever is shorter) would be clearer?

Speaking of, "verify the signature" a couple lines below is completely redundant, the code is pretty clear about that, please just remove that one.

Reply:
fixed